File: 1044-find-common-characters/find-common-characters.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<String> commonChars(String[] words) {
        // 1. Populate the list with ALL characters from the very first word
        List<String> commonList = new ArrayList<>();
        for (char ch : words[0].toCharArray()) {
            commonList.add(String.valueOf(ch));
        }
        
        // 2. Intersect this list with every subsequent word
        for (int i = 1; i < words.length; i++) {
            List<String> currentWordChars = new ArrayList<>();
            for (char ch : words[i].toCharArray()) {
                currentWordChars.add(String.valueOf(ch));
            }
            
            // Temporary list to hold characters common to both commonList and the current word
            List<String> intersection = new ArrayList<>();
            
            for (String s : commonList) {
                // If the current word contains the character, it's a match!
                if (currentWordChars.contains(s)) {
                    intersection.add(s);
                    // Remove it from the current word list so duplicate letters aren't reused incorrectly
                    currentWordChars.remove(s); 
                }
            }
            
            // Update our master list to only contain elements that survived the intersection
            commonList = intersection;
        }
        
        return commonList;
    }
}

File: 1044-find-common-characters/find-common-characters.cpp
class Solution {
public:
    vector<string> commonChars(vector<string>& words) {
        int n=words.size();
        
        vector<int>minfreq(26,INT_MAX);
        for(int i=0;i<n;i++)
        {
            vector<int>temp(26,0);
            for(char ch:words[i])
            {
                temp[ch-'a']++;
            }
            for(int i=0;i<26;i++)
            {
                minfreq[i]=min(minfreq[i],temp[i]);
            }
        }
        vector<string>ans;
        for(int i=0;i<26;i++)
        {
            while(minfreq[i]>0)
            {
                ans.emplace_back(string(1,(char)(i+'a')));
                minfreq[i]--;
            }
        }

        return ans;
    }
};